db/types.go

```diff
@@ -0,0 +1,196 @@
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBWriter is a write-only transaction interface.
+// It is safe for concurrent writes, following an optimistic (OCC) strategy, detecting any write
+// conflicts and returning an error on commit, rather than locking the DB.
+// This can be used to wrap a write-optimized batch object if provided by the backend implementation.
+type DBWriter interface {
+	// Set sets the value for the given key, replacing it if it already exists.
+	// CONTRACT: key, value readonly []byte
+	Set([]byte, []byte) error
+
+	// Delete deletes the key, or does nothing if the key does not exist.
+	// CONTRACT: key readonly []byte
+	Delete([]byte) error
+
+	// Flushes pending writes and discards the transaction.
+	// TODO: maybe change to Flush() and follow WriteBatch semantics (ie. don't discard)
+	Commit() error
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBReadWriter is a transaction interface that allows both reading and writing.
+type DBReadWriter interface {
+	DBReader
+	DBWriter
+}
+
+// Iterator represents an iterator over a domain of keys. Callers must call Close when done.
+// No writes can happen to a domain while there exists an iterator over it, some backends may take
+// out database locks to ensure this will not happen.
+//
+// Callers must make sure the iterator is valid before calling any methods on it, otherwise
+// these methods will panic. This is in part caused by most backend databases using this convention.
+//
+// As with DBReader, keys and values should be considered read-only, and must be copied before they are
+// modified.
+//
+// Typical usage:
+//
+// var itr Iterator = ...
+// defer itr.Close()
+//
+// for ; itr.Valid(); itr.Next() {
+//   k, v := itr.Key(); itr.Value()
+//   ...
+// }
+// if err := itr.Error(); err != nil {
+//   ...
+// }
+type Iterator interface {
+	// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
+	// CONTRACT: start, end readonly []byte
+	Domain() (start []byte, end []byte)
+
+	// Valid returns whether the current iterator is valid. Once invalid, the Iterator remains
+	// invalid forever.
+	Valid() bool
+
+	// Next moves the iterator to the next key in the database, as defined by order of iteration.
+	// If Valid returns false, this method will panic.
+	Next()
```

Comment:
With the `Valid` and `Error` methods, the caller has enough info to avoid a panic, so it seems redundant. And we would also need to add `error` to `Key()` and `Value()`.

Reply:
let's continue he discussion in the `Valid` function.